### PR TITLE
ENH: Add basic driver detection for to_file

### LINKED ISSUE #1607: Use GPKG as the default driver for `to_file`?
Curios what the general thoughts are on switching to use geopackage as the default driver when writing to a file?

http://switchfromshapefile.org/#geopackage

### PR DESCRIPTION
**Pull request by snowman2**
xref #1607

It doesn't support everything, but it does support the main 3 and raises an error when the extension is undetermined. I think the error will be very helpful to prevent https://github.com/geopandas/geopandas/issues/1607#issuecomment-688488076.

### CODE REVIEW

**Comment by martinfleis** [comment]
Thanks! I like the directon, few notes though.

Currently, if you do `df.to_file('name')` without anything else, geopandas creates a folder called `name` and saves shapefile within it. Which is quite handy and I would certainly keep this behaviour.

I think that the only change this PR should bring is to automatically detect extension if one is specified and set the driver accordingly. If none is specified or not recognised, that we should do what we do now, i.e. save shapefile. 

This now brings way too many changes at the same time :).

Also, I don't think this should close #1607 as the discussion on defaults should be done separately from this PR.

**Comment by jorisvandenbossche** [comment]
GDAL has information about the recognized extensions, eg:

```
In [22]: import ogr   

In [23]: driver = ogr.GetDriverByName("GPKG") 

In [24]: driver.GetMetadataItem("DMD_EXTENSION")     
Out[24]: 'gpkg'

In [25]: driver.GetMetadataItem("DMD_EXTENSIONS")   
Out[25]: 'gpkg'

In [26]: driver = ogr.GetDriverByName("GeoJSON")  

In [27]: driver.GetMetadataItem("DMD_EXTENSIONS")    
Out[27]: 'json geojson'
```

but it doesn't seem that `fiona` exposes this (it only uses it under the hood in MemoryFile). 

Long term we should probably also bring it up to move (part of) this functionality to fiona.

**Comment by snowman2** [comment]
> Currently, if you do df.to_file('name') without anything else, geopandas creates a folder called name and saves shapefile within it. Which is quite handy and I would certainly keep this behaviour.

Just added a fix for this.

> I don't think this should close #1607 as the discussion on defaults should be done separately from this PR.

Actually, I think they are tied together. The behavior of just defaulting to the shapefile driver without error has thrown myself and many others off when forgetting to explicitly set the driver. I have seen this so many times. Because of this, I think it would be highly beneficial to users to see an exception if the driver cannot be found by default.

> Long term we should probably also bring it up to move (part of) this functionality to fiona.

Agreed.

**Comment by snowman2** [comment]
> Long term we should probably also bring it up to move (part of) this functionality to fiona.

https://github.com/Toblerity/Fiona/pull/948

**Comment by snowman2** [comment]
> If none is specified or not recognised, that we should do what we do now, i.e. save shapefile.

Would it be acceptable for you to default to `ESRI Shapefile` with a warning in this scenario?

Maybe something like:

> Could not determine the driver based on your extension {extension}. Defaulting to the ESRI Shapefile driver. In a future release this will raise an exception.

**Comment by snowman2** [comment]
Modified this to pass None to Fiona if the driver is not found. The thinking behind this is to enable better driver detection if https://github.com/Toblerity/Fiona/pull/948 is added. Thoughts?

File: geopandas/io/file.py
Comment (martinfleis) [inline geopandas/io/file.py:223]: I think this is fine if we use the complete dict I proposed above. I would maybe add a warning here explaining how to fix the issue because fiona's error shows `DriverError: no driver` only.

File: geopandas/geodataframe.py
Comment (martinfleis) [inline geopandas/geodataframe.py:714]:
```suggestion
            If not specified, it attempts to infer it from the file extension.
            If no extension is specified, it saves ESRI Shapefile to a folder.
```

File: geopandas/geoseries.py
Comment (martinfleis) [inline geopandas/geoseries.py:253]:
```suggestion
            If not specified, it attempts to infer it from the file extension.
            If no extension is specified, it saves ESRI Shapefile to a folder.
```

File: geopandas/io/file.py
Comment (martinfleis) [inline geopandas/io/file.py:243]:
```suggestion
        If not specified, it attempts to infer it from the file extension.
        If no extension is specified, it saves ESRI Shapefile to a folder.
```

**Review by martinfleis — COMMENTED** [review]
> What are the current blockers to merge this?

Time :).

Thanks a lot for working on this. I am not 100% sure about all changes we do here, since it may break some things (e.g. unrecognised suffix is currently saved as shapefile folder, which breaks now), but generally, it looks good.

File: geopandas/io/file.py
Comment (snowman2) [reply geopandas/io/file.py:207]: I would much prefer the hardcoded version. I would to take manually adding the driver any day to adding `ogr` as a dependency 😀.

File: geopandas/io/file.py
Comment (snowman2) [reply geopandas/io/file.py:223]: Sounds like a good idea 👍

File: geopandas/io/file.py
Comment (martinfleis) [reply geopandas/io/file.py:207]: Yeah, I missed "if we find a way how to do that in fiona" above. We may write a test checking if the dict is up-do-date.

Comment (jorisvandenbossche) [reply geopandas/io/file.py:207]: Yes, we probably don't want to `import ogr`, it is know to be problematic together with `import fiona` from time to time, AFAIR

**Comment by snowman2** [comment]
I updated it to attempt to get the driver extension map from https://github.com/Toblerity/Fiona/pull/948 if it exists first and use the other map as a backup (so probably want to wait until that is merged first). That should keep it up-to-date. I also added the `ESRI Shapefile` driver as the default for minimal breakage to existing behavior. This is still problematic in case you have a typo in the extension (e.g. `gpkp`), but it doesn't break the folder behavior. I think raising an exception in the future may be a good idea when the driver is not found if the folder extension(s) supported are better defined. Thoughts?

**Comment by martinfleis** [comment]
> I think raising an exception in the future may be a good idea when the driver is not found if the folder extension(s) supported are better defined.

Agree. Can you add a comment into the code to remind us?

> I updated it to attempt to get the driver extension map from Toblerity/Fiona#948 if it exists first and use the other map as a backup (so probably want to wait until that is merged first)

That doesn't seem to go any further. We should maybe ignore that and rely on our own mapping dict. It will never be perfect but it'll cover most of use cases. Any other can be manually specified.

The failure is probably related to #1986.

**Review by martinfleis — APPROVED** [review]
Thanks!

File: geopandas/io/tests/test_file.py
Comment (jorisvandenbossche) [inline geopandas/io/tests/test_file.py:88]: Do we check somewhere that we actually write the correct format? Because eg in this test we are reading it back, but that will also pass if we were still always writing shapefiles (if the detection would not have worked)

**Review by jorisvandenbossche — COMMENTED** [review]
Let's try to get this in for the upcoming release! 

Looks good, just one question about the tests

File: geopandas/io/tests/test_file.py
Comment (snowman2) [reply geopandas/io/tests/test_file.py:95]: Sounds like a good idea to me. I added `assert_correct_driver` to check.

**Review by jorisvandenbossche — APPROVED** [review]
Thanks for the update!

**Comment by snowman2** [comment]
Thanks for the reviews 👍

**Pull request merged**
